### ISSUE TITLE
Assertion failure in operator delete of llvm::User

### ISSUE THREAD

**Issue by marcauberer** [llvm/llvm-project#129900]
Commit https://github.com/llvm/llvm-project/commit/e3f936eb755d9ae37019ffcc7f53d71d2d58d188 introduces a functional regression. If the constructor after operator new fails, and then operator delete is called, we run into the following assertion. This was caught by malfunction tests:

https://github.com/llvm/llvm-project/blob/main/llvm/lib/IR/User.cpp#L190

Reverting the removal of:

```cpp
Obj->NumUserOperands = Us;
Obj->HasHungOffUses = false;
Obj->HasDescriptor = DescBytes != 0;
```
respecively
```cpp
Obj->NumUserOperands = 0;
Obj->HasHungOffUses = true;
Obj->HasDescriptor = false;
```
helps.

I will prepare a patch for that.
